daemons/ipa-kdb/ipa_kdb_mspac.c

```diff
@@ -1827,37 +1827,57 @@ krb5_error_code filter_logon_info(krb5_context context,
     bool result;
     char *domstr = NULL;
 
-    domain = get_domain_from_realm_update(context, realm);
-    if (!domain) {
-        return EINVAL;
+    ipactx = ipadb_get_context(context);
+    if (!ipactx || !ipactx->mspac) {
+        return KRB5_KDB_DBNOTINITED;
     }
 
-    /* check netbios/flat name */
-    if (strcasecmp(info->info->info3.base.logon_domain.string,
-                   domain->flat_name) != 0) {
-        krb5_klog_syslog(LOG_ERR, "PAC Info mismatch: domain = %s, "
-                                  "expected flat name = %s, "
-                                  "found logon name = %s",
-                                  domain->domain_name, domain->flat_name,
-                                  info->info->info3.base.logon_domain.string);
+    domain = get_domain_from_realm_update(context, realm);
+    if (!domain) {
         return EINVAL;
     }
 
     /* check exact sid */
     result = dom_sid_check(&domain->domsid, info->info->info3.base.domain_sid, true);
     if (!result) {
-        domstr = dom_sid_string(NULL, info->info->info3.base.domain_sid);
-        if (!domstr) {
+        struct ipadb_mspac *mspac_ctx = ipactx->mspac;
+        result = FALSE;
+        /* Didn't match but perhaps the original PAC was issued by a child domain's DC? */
+        for (k = 0; k < mspac_ctx->num_trusts; k++) {
+            result = dom_sid_check(&mspac_ctx->trusts[k].domsid,
+                             info->info->info3.base.domain_sid, true);
+            if (result) {
+                domain = &mspac_ctx->trusts[k];
+                break;
+            }
+        }
+        if (!result) {
+            domstr = dom_sid_string(NULL, info->info->info3.base.domain_sid);
+            if (!domstr) {
+                return EINVAL;
+            }
+            krb5_klog_syslog(LOG_ERR, "PAC Info mismatch: domain = %s, "
+                                    "expected domain SID = %s, "
+                                    "found domain SID = %s",
+                                    domain->domain_name, domain->domain_sid, domstr);
+            talloc_free(domstr);
             return EINVAL;
         }
+    }
```

Comment:
I'm not a fan of multiple exit points. Is there a reason that this is not "if (domstr) { /* do report */ }; return EINVAL;" ?

Reply:
In other places in the code where there no sid value it does something like:
```
domstr ? domstr : "<failed to display>
```
It looks like talloc_free can take a NULL fine so this can be reduced to a single return that way.

Reply:
I unified the return statements and followed @rcritten's proposal.